src/counters.ts

```diff
@@ -0,0 +1,701 @@
+		if (this.dumpToFile) {
+			promises.push(this.writeCollectionData());
+			promises.push(writeFileAsync(`${this.getOutputFileNameForProcess()}_processInfo.json`, jsonDump(this.processesToTrack)));
+		}
+		if (this.dumpToChart) {
+			promises.push(this.writeCharts());
+		}
+		await Promise.all(promises);
+	}
+
+	private async writeCharts(): Promise<void> {
+		const trace = require('debug')(`${logPrefix}:writeCharts:trace`);
+		let promises: Promise<void>[] = [];
+		[Counters.WholeComputerProcessInfo, Counters.TotalsProcessInfo, ...this.processesToTrack].forEach((proc: ProcessInfo) => {
+			let file = `${this.getOutputFileNameForProcess(proc)}_chart.png`;
+			trace(`chart file name: ${file}`);
+			if (this.collection[proc.pid]) {
+				// if we have collected data for this process then write Chart for it.
+				promises.push(this.writeChart(file, this.collection[proc.pid]));
+			}
+			if (this.includeMovingAverages) {
+				file = `${this.getOutputFileNameForProcess(proc)}_sma_chart.png`;
+				trace(`sma chart file name: ${file}`);
+				if (this.smaOver4Collection[proc.pid]) {
+					// if we have simple moving average data computed for this process then  write Chart for it.
+					promises.push(this.writeChart(file, this.smaOver4Collection[proc.pid]));
+				}
+				file = `${this.getOutputFileNameForProcess(proc)}_ema_chart.png`;
+				trace(`ema chart file name: ${file}`);
+				if (this.emaOver4Collection[proc.pid]) {
+					// if we have exponential moving average data computed for this process then  write Chart for it.	
+					promises.push(this.writeChart(file, this.emaOver4Collection[proc.pid]));
+				}
+			}
+		});
+		trace(`waiting for all chart files to be written out ...`);
+		await Promise.all(promises);
+	}
+
+	private async writeChart(file: string, processCollection: ProcessStatisticsCollection): Promise<void> {
+		const trace = require('debug')(`${logPrefix}:writeChart:trace`);
+		trace(`processStats=${jsonDump(processCollection)}`);
+		const xKey = 'timestamp';
+		const xData: number[] = processCollection[xKey];
+		const xAxisLabel = `${xKey}(${ProcessStatisticsUnits[xKey]})`;
+		const lines: LineData[] = [];
+		const title = path.parse(file).name; //get just file name without directory paths and extension
+		Object.keys(processCollection)
+			.filter(key => key !== 'elapsed' && key !== 'timestamp' && key !== 'pid' && key !== 'ppid')
+			.forEach(key => {
+				lines.push({
+					label: `${key}(${ProcessStatisticsUnits[key]})`,
+					data: processCollection[key]
+				});
+			});
+		trace(`xData:${jsonDump(xData)}, lines:${jsonDump(lines)}, xAxisLabel: ${xAxisLabel}, file:${file}, title:${title}`);
+		await writeChartToFile(xData, lines, 'png', processCollection.timestamp[0], xAxisLabel, file, title);
+	}
+
+	private async writeCollectionData(): Promise<void> {
+		let promises: Promise<void>[] = [];
+			let file = `${this.getOutputFileName()}_data.json`;
+		promises.push(writeFileAsync(file, jsonDump(this.collection)));

```

Comment:
If all tasks are writing to same file in async way, they can be parallel right? Is it not giving intermitten issues like lock/overwrite etc? 
Can we await them one by one? Since these are used in test, I would say  consistency might be more important than parallelization.

Reply:
All writes are to different files by design.

Reply:
There is a lot of data that can be written out for a long run and since the operations can be io intensive it is good to do them in parallel as lab time can be valuable, and logic here is quite straightforward so long as the files being written are all different.

Adding comment to clarify.